Add m.login.appservice to the spec

Half-Shot opened the pull request:
https://github.com/matrix-org/matrix-doc/pull/2778

richvdh reviewed (changes requested):
I think you've put all this in the wrong section. MSC2778 doesn't specify a UIA authentication type, but rather a modification to the `/login` endpoint (which doesn't use UIA). It should be in this section: https://spec.matrix.org/unstable/client-server-api/#login.

Half-Shot commented:
Sorry, my mistake. I homed in on the thing that looked like login types without thinking.

I've moved and reworked the section to hopefully make it fit better with `/login`.

richvdh commented on content/client-server-api/_index.md:
```diff
@@ -1021,10 +1021,46 @@ client supports it, the client should redirect the user to the
+
+{{% boxes/note %}}
+Appservices do not need to login as individual users in all cases, as they
+can perform [Identity Assertion](http://localhost:1313/application-service-api/#identity-assertion)
```
link looks wrong

uhoreg replied:
```suggestion
can perform [Identity Assertion](../application-service-api#identity-assertion)
```

richvdh commented on content/client-server-api/_index.md:
```diff
@@ -1021,10 +1021,46 @@ client supports it, the client should redirect the user to the
 
 {{% http-api spec="client-server" api="logout" %}}
 
+
```
```suggestion
```

richvdh commented on content/client-server-api/_index.md:
```diff
@@ -1021,10 +1021,46 @@ client supports it, the client should redirect the user to the
+or the user has not previously been registered then the homeserver will
+respond with an errcode of `M_FORBIDDEN`.
+
+If the access token does correspond to an appservice, but the user does
+not exist within it's namespace then the homeserver will respond with an
```
```suggestion
not exist within its namespace then the homeserver will respond with an
```

richvdh replied:
```suggestion
If the access token does correspond to an appservice, but the user id does
not lie within it's namespace then the homeserver will respond with an
```

richvdh commented on changelogs/client_server/newsfragments/3324.feature:
```diff
@@ -0,0 +1 @@
+Add support for m.login.appservice, as per [MSC2778](https://github.com/matrix-org/matrix-doc/pull/2778).
```
```suggestion
Extend `/_matrix/client/r0/login` to accept a `m.login.appservice`, as per [MSC2778](https://github.com/matrix-org/matrix-doc/pull/2778).
```

richvdh reviewed (changes requested):
Thanks, looks better now. A few nits.

Would it be helpful to add a link from the appservice spec?

turt2live commented:
@Half-Shot do you have bandwidth to update this?

Half-Shot commented:
My bandwidth is in the negatives atm

turt2live commented:
Would you be opposed to this being taken over to get it over the line?

uhoreg commented on content/client-server-api/_index.md:
```diff
@@ -1021,10 +1021,46 @@ client supports it, the client should redirect the user to the
 
+#### Appservice Login
+
+A appservice can login by providing a valid appservice token and a user within the appservice's
```
```suggestion
An appservice can log in by providing a valid appservice token and a user within the appservice's
```

uhoreg commented on content/client-server-api/_index.md:
```diff
@@ -1021,10 +1021,46 @@ client supports it, the client should redirect the user to the
+namespace. 
+
+{{% boxes/note %}}
+Appservices do not need to login as individual users in all cases, as they
```
```suggestion
Appservices do not need to log in as individual users in all cases, as they
```

turt2live commented on content/client-server-api/_index.md:
```diff
@@ -1021,6 +1021,41 @@ client supports it, the client should redirect the user to the
+for a single user then they can use this API instead.
+{{% /boxes/note %}}
+
+This request must be authenticated by the [appservice `as_token`](../application-service-api#registration) 
```
```suggestion
This request must be authenticated by the [appservice `as_token`](/application-service-api#registration) 
```

turt2live reviewed (approved):
otherwise lgtm - thanks!

uhoreg commented on content/client-server-api/_index.md:
```diff
@@ -1021,6 +1021,41 @@ client supports it, the client should redirect the user to the
+
+{{% boxes/note %}}
+Appservices do not need to log in as individual users in all cases, as they
+can perform [Identity Assertion](../application-service-api#identity-assertion)
```
```suggestion
can perform [Identity Assertion](/application-service-api#identity-assertion)
```
